### PR TITLE
Validate if products have at least 1 quantity

### PR DESCRIPTION
**Pull request by milwoz**

### CODE REVIEW

File: src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php [hunk 1]
```diff
@@ -79,6 +83,16 @@ private function isInStock(WishlistItem $wishlistProduct): bool
         return false;
     }
 
+    private function productHasPositiveQuantity(OrderItemInterface $product): bool
```
Comment (senghe) [inline src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php:86]: Could you extract it into checker class and spec?

File: src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php [hunk 2]
```diff
@@ -49,22 +50,25 @@ public function __construct(
...
-                continue;
+            if ($this->productCanBeProcessed($wishlistProduct)) {
+                $this->addProductToWishlist($wishlistProduct);
             }
-
-            $this->addProductToWishlist($wishlistProduct);
         }
     }
 
+    private function productCanBeProcessed(WishlistItem $wishlistProduct): bool
```
Comment (senghe) [inline src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php:65]: Could you extract this into separate checker and spec it?

**Review by senghe — CHANGES_REQUESTED** [review]

File: src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php (hunk 2 above)
Comment (milwoz) [reply src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php]: Done.

File: src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php (hunk 1 above)
Comment (milwoz) [reply src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php]: Done.

File: spec/Checker/ProductQuantityCheckerSpec.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    public function it_has_zero_products(
+        OrderItemInterface $product,
+        TranslatorInterface $translator,
+        FlashBagInterface $flashBag
+    ): void {
+        $product->getQuantity()->willReturn(0);
+
+        $translator->trans('bitbag_sylius_wishlist_plugin.ui.increase_quantity')->willReturn('Increase the quantity of at least one item.');
+
+        $flashBag->add('error', 'Increase the quantity of at least one item.')->shouldBeCalled();
```
Comment (senghe) [inline spec/Checker/ProductQuantityCheckerSpec.php:54]: The flashbag is a part of HTTP controllers logic. We have to move it into controller and adjust the logic in the code to make it happen.

File: spec/Checker/ProductCanBeProcessedCheckerSpec.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+use BitBag\SyliusWishlistPlugin\Checker\ProductCanBeProcessedChecker;
+use BitBag\SyliusWishlistPlugin\Checker\ProductQuantityCheckerInterface;
+use BitBag\SyliusWishlistPlugin\Command\Wishlist\WishlistItem;
+use PhpSpec\ObjectBehavior;
+use Sylius\Bundle\OrderBundle\Controller\AddToCartCommandInterface;
+use Sylius\Component\Core\Model\OrderItemInterface;
+use Symfony\Component\HttpFoundation\Session\Flash\FlashBagInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class ProductCanBeProcessedCheckerSpec extends ObjectBehavior
```
Comment (senghe) [inline spec/Checker/ProductCanBeProcessedCheckerSpec.php:23]: Finals in all classes 🙏 and there is a missing scenario 🙃

File: src/Checker/ProductCanBeProcessedChecker.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        $cartItem = $wishlistProduct->getCartItem()->getCartItem();
+
+        return $this->isInStock($wishlistProduct) && $this->productQuantityChecker->productHasPositiveQuantity($cartItem);
+    }
+
+    private function isInStock(WishlistItem $wishlistProduct): bool
+    {
+        $cartItem = $wishlistProduct->getCartItem()->getCartItem();
+
+        if (0 < $wishlistProduct->getCartItem()->getCartItem()->getQuantity()) {
```
Comment (senghe) [inline src/Checker/ProductCanBeProcessedChecker.php:46]: Could you use Demeter law here?

File: src/Checker/ProductCanBeProcessedChecker.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusWishlistPlugin\Checker;
+
+use BitBag\SyliusWishlistPlugin\Command\Wishlist\WishlistItem;
+use Symfony\Component\HttpFoundation\Session\Flash\FlashBagInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class ProductCanBeProcessedChecker implements ProductCanBeProcessedCheckerInterface
```
Comment (senghe) [inline src/Checker/ProductCanBeProcessedChecker.php:17]:
```suggestion
class ProductProcessingChecker implements ProductCanBeProcessedCheckerInterface
```

File: src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php [hunk 7]
```diff
@@ -90,5 +78,10 @@ private function addProductToWishlist(WishlistItem $wishlistProduct): void
 
         $this->orderModifier->addToOrder($cart, $cartItem);
         $this->orderRepository->add($cart);
+
+        if (false === $this->flashBag->has('success')){
+            $this->flashBag->add('success', $this->translator->trans('bitbag_sylius_wishlist_plugin.ui.added_to_cart'));
```
Comment (senghe) [inline src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php:83]: We should throw an exception here and move flashbag to a controller.

**Review by senghe — CHANGES_REQUESTED** [review]

File: spec/Checker/ProductQuantityCheckerSpec.php (hunk 3 above)
Comment (milwoz) [reply spec/Checker/ProductQuantityCheckerSpec.php:54]: Done.

File: src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php (hunk 7 above)
Comment (milwoz) [reply src/CommandHandler/Wishlist/AddSelectedProductsToCartHandler.php:83]: Done.

File: src/Checker/ProductCanBeProcessedChecker.php (hunk 6 above)
Comment (milwoz) [reply src/Checker/ProductCanBeProcessedChecker.php:17]: Done.

File: spec/Checker/ProductCanBeProcessedCheckerSpec.php (hunk 4 above)
Comment (milwoz) [reply spec/Checker/ProductCanBeProcessedCheckerSpec.php:23]: Done/

File: src/Checker/ProductCanBeProcessedChecker.php (hunk 5 above)
Comment (milwoz) [reply src/Checker/ProductCanBeProcessedChecker.php]: Done.

**Review by senghe — APPROVED** [review]

**Pull request merged**